## 1. About task and Data Set Consideration: 

Task: Digit classification using Convolutional Neural Network. 
Data-Set: MNIST.
Previous-Result: With two Densly connected layer, the accuracy: 97.8%


### 1.1. Dependencies:


In [1]:
import numpy as np
import matplotlib.pyplot as plt
import keras
from keras import models
from keras import layers
from keras.datasets import mnist
from keras.utils import to_categorical

keras.__version__


C:\Users\home\Anaconda3\lib\site-packages\h5py\__init__.py:34: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


'2.1.3'

### 1.2. Initial Explorations: 

In [2]:
model = models.Sequential()
model.add(layers.Conv2D(32,(3,3),activation='relu',input_shape=(28,28,1)))
model.add(layers.MaxPooling2D((2,2)))
model.add(layers.Conv2D(64,(3,3),activation='relu'))
model.add(layers.MaxPooling2D((2,2)))
model.add(layers.Conv2D(64,(3,3),activation='relu'))

model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 3, 3, 64)          36928     
Total params: 55,744
Trainable params: 55,744
Non-trainable params: 0
_________________________________________________________________


### 1.3. Adding a classifier on the top of Convnet:


In [3]:
model.add(layers.Flatten())
model.add(layers.Dense(64,activation='relu'))
model.add(layers.Dense(10,activation='softmax'))
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 3, 3, 64)          36928     
_________________________________________________________________
flatten_1 (Flatten)          (None, 576)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 64)                36928     
__________

 ## 2. Loading data:

In [4]:
(train_images,train_labels),(test_images,test_labels) = mnist.load_data()


## 3. Preparing Data:


In [5]:
train_images = train_images.reshape((60000,28,28,1))
train_images = train_images.astype('float32') / 255
test_images = test_images.reshape((10000,28,28,1))
test_images = test_images.astype('float32') / 255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)


## 4. Building Network:


In [6]:
model.compile(optimizer='rmsprop',loss='categorical_crossentropy',metrics=['acc'])
model.fit(train_images,train_labels,epochs=5,batch_size=64)


Epoch 1/5
60000/60000 [==============================] - 105s 2ms/step - loss: 0.1735 - acc: 0.9459
Epoch 2/5
60000/60000 [==============================] - 103s 2ms/step - loss: 0.0483 - acc: 0.9854
Epoch 3/5
60000/60000 [==============================] - 104s 2ms/step - loss: 0.0335 - acc: 0.9896
Epoch 4/5
60000/60000 [==============================] - 103s 2ms/step - loss: 0.0253 - acc: 0.9926
Epoch 5/5
60000/60000 [==============================] - 108s 2ms/step - loss: 0.0199 - acc: 0.9941


## 5. Model Evaluation on Test Set:

In [7]:
test_loss,test_accuracy = model.evaluate(test_images,test_labels)
test_accuracy

10000/10000 [==============================] - 6s 619us/step


0.9921

## 6. Conclusion:

1. Densely Connected Network, accuracy: 97.8%.
2. Basic Convolutional Network, Test accuracy: 99.3%.
3. Decrease Error rate, relatively by: 68%.
    